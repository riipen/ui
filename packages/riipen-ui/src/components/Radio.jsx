import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";

import Typography from "./Typography";

class Radio extends React.Component {
  static propTypes = {
    /**
     * If `true`, the component is checked.
     */
    checked: PropTypes.bool,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf(["primary", "secondary", "default"]),

    /**
     * If `true`, the radio will be disabled.
     */
    disabled: PropTypes.bool,

    /**
     * Label text to display for the radio.
     */
    label: PropTypes.string
  };

  static defaultProps = {
    checked: false,
    disabled: false
  };

  static contextType = ThemeContext;

  render() {
    const { checked, classes, color, disabled, label, ...other } = this.props;

    const theme = this.context;

    return (
      <div className={clsx(classes)}>
        <label htmlFor={other.id}>
          <Typography>{label}</Typography>
          <input
            checked={checked}
            disabled={disabled}
            type="radio"
            {...other}
          />
          <span
            className={clsx("checkmark", disabled ? "disabled" : null, color)}
          />
        </label>
        <style jsx>{`
          label {
            cursor: pointer;
            display: block;
            margin-bottom: 12px;
            padding-left: 35px;
            position: relative;
            user-select: none;
          }

          /* Hide the browser's default checkbox */
          label input {
            cursor: pointer;
            opacity: 0;
            position: absolute;
          }

          /* Create a custom checkbox */
          .checkmark {
            background-color: #eee;
            border-radius: 50%;
            border: 1px solid rgba(0, 0, 0, 0.23);
            height: 20px;
            left: 0;
            position: absolute;
            top: 0;
            width: 20px;
          }

          /* When the checkbox is checked */
          label input:checked ~ .checkmark {
            background-color: ${theme.palette.grey[400]};
            border-color: ${theme.palette.grey[400]};
          }

          label input:checked ~ .primary {
            background-color: ${theme.palette.primary.main};
            border-color: ${theme.palette.primary.main};
          }
          label input:checked ~ .secondary {
            background-color: ${theme.palette.secondary.main};
            border-color: ${theme.palette.secondary.main};
          }

          /* Create the checkmark/indicator (hidden when not checked) */
          .checkmark:after {
            content: "";
            display: none;
            position: absolute;
          }

          /* Show the checkmark when checked */
          label input:checked ~ .checkmark:after {
            display: block;
          }

          /* Style the checkmark/indicator */
          label .checkmark:after {
            background: white;
            border-radius: 50%;
            height: 10px;
            left: 5px;
            top: 5px;
            width: 10px;
          }

          .disabled {
            background-color: ${theme.palette.grey[300]};
            border-color: ${theme.palette.grey[300]};
            pointer-events: none;
          }
        `}</style>
      </div>
    );
  }
}

export default Radio;
