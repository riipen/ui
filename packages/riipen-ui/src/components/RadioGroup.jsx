import PropTypes from "prop-types";
import React from "react";

import withClasses from "../utils/withClasses";

import InputLabel from "./InputLabel";
import InputHint from "./InputHint";
import Typography from "./Typography";

class RadioGroup extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An error to display below the input.
     */
    error: PropTypes.node,

    /**
     * Hint text to display under the label of the input.
     */
    hint: PropTypes.node,

    /**
     * Label text to display for the input.
     */
    label: PropTypes.node,

    /**
     * Callback fired when a radio button is selected.
     *
     * @param {object} event The event source of the callback.
     * You can pull out the new value by accessing `event.target.value` (string).
     */
    onChange: PropTypes.func,

    /**
     * If true, an asterisk will be appended to the end of the label.
     */
    required: PropTypes.bool,

    /**
     * A warning to display below the input.
     */
    warning: PropTypes.node
  };

  handleChange = e => {
    const { onChange } = this.props;

    if (onChange) {
      onChange(e, e.target.value);
    }
  };

  render() {
    const {
      children,
      error,
      hint,
      label,
      required,
      warning,
      ...other
    } = this.props;

    const childrenWithProps = React.Children.map(children, child =>
      React.cloneElement(child, { onChange: this.handleChange })
    );

    return (
      <React.Fragment>
        <fieldset>
          {label && (
            <InputLabel htmlFor={other.name} required={required}>
              {label}
            </InputLabel>
          )}
          {hint && <InputHint>{hint}</InputHint>}
          {childrenWithProps}
          {error && (
            <Typography color="negative" variant="body2">
              {error}
            </Typography>
          )}
          {warning && (
            <Typography color="secondary" variant="body2">
              {warning}
            </Typography>
          )}
        </fieldset>
        <style jsx>{`
          fieldset {
            border: none;
            margin: 0;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(RadioGroup);
