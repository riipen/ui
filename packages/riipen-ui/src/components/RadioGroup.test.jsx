import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import RadioGroup from "./RadioGroup.jsx";
import Radio from "./Radio.jsx";

describe("<RadioGroup>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<RadioGroup />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<RadioGroup />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("renders given non null children", () => {
      const child = <Radio label="one" readOnly />;
      const child2 = <Radio label="two" readOnly />;

      const wrapper = mount(
        <RadioGroup>
          {child}
          {null}
          {child2}
        </RadioGroup>
      );

      const childrenNodes = wrapper.find("fieldset").find("Radio");
      expect(childrenNodes.at(0).props().label).toEqual("one");
      expect(childrenNodes.at(1).props().label).toEqual("two");
    });
  });

  describe("error prop", () => {
    it("displays error", () => {
      const error = <span>an error!</span>;

      const wrapper = mount(<RadioGroup error={error} />);

      expect(wrapper.find("fieldset").contains(error)).toEqual(true);
    });
  });

  describe("hint prop", () => {
    it("renders InputLabel when hint is provided", () => {
      const hint = "it's a hint";

      const wrapper = mount(<RadioGroup hint={hint} />);

      expect(wrapper.find("InputLabel")).toHaveLength(1);
    });

    it("passes given hint to InputLabel", () => {
      const hint = "it's a hint";

      const wrapper = mount(<RadioGroup hint={hint} />);

      expect(wrapper.find("InputLabel").props().hint).toEqual(hint);
    });
  });

  describe("label prop", () => {
    it("renders Inputlabel and passes label to InputLabel as children", () => {
      const label = "it's a label";

      const wrapper = mount(<RadioGroup label={label} />);

      expect(wrapper.find("InputLabel").props().children).toEqual(label);
      expect(wrapper.find("InputLabel")).toHaveLength(1);
    });

    it("does not render InputLabel when both label and hint are not provided", () => {
      const wrapper = mount(<RadioGroup />);

      expect(wrapper.find("InputLabel")).toHaveLength(0);
    });
  });

  describe("name prop", () => {
    it("passes given name to InputLabel as htmlFor prop", () => {
      const name = "hello";
      const label = "it's a label";
      const child = <Radio id="one" />;

      const wrapper = mount(
        <RadioGroup name={name} label={label}>
          {child}
        </RadioGroup>
      );

      expect(wrapper.find("InputLabel").props().htmlFor).toEqual(name);
    });
  });

  describe("onChange prop", () => {
    it("sets onChange prop of children nodes", () => {
      const child = <Radio label="one" readOnly />;

      const wrapper = mount(<RadioGroup>{child}</RadioGroup>);

      expect(wrapper.find("Radio").props().onChange).toBeDefined();
    });

    it("invokes handler when child Radio is changed", () => {
      const child = <Radio label="one" readOnly />;
      const onChange = jest.fn();

      const wrapper = mount(
        <RadioGroup onChange={onChange}>{child}</RadioGroup>
      );

      wrapper
        .find("Radio")
        .find("input")
        .invoke("onChange")({ target: { value: "a" } });

      expect(onChange).toHaveBeenCalledWith({ target: { value: "a" } }, "a");
    });
  });

  describe("required prop", () => {
    it("passes required prop to InputLabel", () => {
      const label = "it's a label";
      const required = true;

      const wrapper = mount(<RadioGroup label={label} required={required} />);

      expect(wrapper.find("InputLabel").props().required).toEqual(required);
    });
  });

  describe("suffix prop", () => {
    it("passes suffix prop to InputLabel", () => {
      const suffix = <span>"Test"</span>;

      const wrapper = mount(<RadioGroup suffix={suffix} />);

      expect(wrapper.find("InputLabel").props().suffix).toEqual(suffix);
    });
  });

  describe("value prop", () => {
    it("sets the childs checked prop according to the value", () => {
      const child = <Radio value="one" id="one" />;

      const wrapper = mount(<RadioGroup value="one">{child}</RadioGroup>);

      expect(wrapper.find("Radio").props().checked).toEqual(true);
    });
  });

  describe("warning prop", () => {
    it("displays warning", () => {
      const warning = "oh no";

      const wrapper = mount(<RadioGroup warning={warning} />);

      expect(wrapper.find("fieldset").contains(warning)).toEqual(true);
    });
  });
});
