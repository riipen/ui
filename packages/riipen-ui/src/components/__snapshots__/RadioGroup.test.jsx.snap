// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup> renders correct snapshot 1`] = `
<withClasses(RadioGroup)>
  <RadioGroup
    classes={
      Array [
        "riipen",
        "riipen-radiogroup",
      ]
    }
  >
    <fieldset
      className="jsx-3241897731"
    />
    <JSXStyle
      id="3241897731"
    />
  </RadioGroup>
</withClasses(RadioGroup)>
`;
